### PR TITLE
CI/Simulator: Drive a loop in metadrive

### LINKED ISSUE #30694: [$500 bounty] Drive a loop in MetaDrive
Extend the MetaDrive test to drive a defined course and ensure it stays in the line. Requirements:
* at least 1 minute of driving
* passes reliably: 20+ times in a row
* test fails if a reasonable user would have disengaged during the loop (drives out of lane, etc.)

Ideally, we can also configure the distance and/or runtime of the test with a simple parameter so that we can run a quick one in CI, and perhaps a longer version on master.

https://github.com/commaai/openpilot/blob/master/tools/sim/tests/test_sim_bridge.py

### PR DESCRIPTION
**Pull request by bongbui321**
Resolve [#30694](https://github.com/commaai/openpilot/issues/30694)

~Currently, the simulator will fail when it tries to cross lane , and can change the track_size for CI testing. can test by `./test_metadrive_bridge.py 30`, which would be a fail case on the curve~

Try and run test locally with this cmd: `pytest tools/sim/tests/test_metadrive_bridge.py --time_done=<seconds>`

~Depends on: https://github.com/metadriverse/metadrive/pull/710~

- [x] deterministic test
~- [ ] update poetry~ Metadrive maintainer is a while to review, detect out of lane locally for now
~- [ ] maybe more test cases (lead car, etc)~ traffic is extremely expensive even running locally with decent gpu

### CODE REVIEW

**Comment by adeebshihadeh** [comment]
>   ~- [ ] maybe more test cases (lead car, etc)~ traffic is extremely expensive even running locally with decent gpu

A super reliable test that only tests one thing is way more valuable than a fancy one that's flaky :)

**Comment by bongbui321** [comment]
Tried running locally with `pytest tools/sim/tests/test_metadrive_bridge.py --time_done=600`. can modify to run the loop as long as you want it to be

File: .github/workflows/tools_tests.yaml [hunk 1]
```diff
@@ -76,7 +76,7 @@ jobs:
         ${{ env.RUN }} "export MAPBOX_TOKEN='pk.eyJ1Ijoiam5ld2IiLCJhIjoiY2xxNW8zZXprMGw1ZzJwbzZneHd2NHljbSJ9.gV7VPRfbXFetD-1OVF0XZg' && \
                         source selfdrive/test/setup_xvfb.sh && \
                         source selfdrive/test/setup_vsound.sh && \
-                        CI=1 pytest tools/sim/tests/test_metadrive_bridge.py -W ignore::pyopencl.CompilerWarning"
+                        CI=1 pytest tools/sim/tests/test_metadrive_bridge.py --time_done=90 -W ignore::pyopencl.CompilerWarning"
```
Comment (adeebshihadeh) [inline .github/workflows/tools_tests.yaml:79]: needed to add "-W ignore::pyopencl.CompilerWarning" when running locally as well. let's put this directly in the code so it doesn't need to be specified on the command line (or fix the warning)

File: tools/sim/tests/conftest.py [hunk 2]
```diff
@@ -0,0 +1,8 @@
+import pytest
+
+def pytest_addoption(parser):
+  parser.addoption("--time_done", action="store", default=60, type=int, help="Seconds to run metadrive drive")
```
Comment (adeebshihadeh) [inline tools/sim/tests/conftest.py:4]: call this "test_duration"

File: tools/sim/tests/test_metadrive_bridge.py [hunk 3]
```diff
@@ -9,5 +10,9 @@
 
 @pytest.mark.slow
 class TestMetaDriveBridge(TestSimBridgeBase):
+  @pytest.fixture(autouse=True)
```
Comment (adeebshihadeh) [inline tools/sim/tests/test_metadrive_bridge.py:13]: is this decorator necessary for setup_class?

File: tools/sim/bridge/metadrive/metadrive_process.py [hunk 4]
```diff
@@ -111,10 +112,18 @@ def get_cam_as_rgb(cam):
         reset()
 
     if rk.frame % 5 == 0:
-      obs, _, terminated, _, info = env.step(vc)
+      _, _, terminated, _, _ = env.step(vc)
+      time_out = True if time.monotonic() - start_time >= time_done else False
+      out_of_lane = env.vehicle.on_broken_line or env.vehicle.on_yellow_continuous_line or env.vehicle.on_white_continuous_line or env.vehicle.crash_sidewalk
```
Comment (adeebshihadeh) [inline tools/sim/bridge/metadrive/metadrive_process.py:117]: any way to check "vehicle in lane" or another simple check? looks like we might have to keep adding conditions for this one

**Review by adeebshihadeh — CHANGES_REQUESTED** [review]
Took me a few attempts to get this running locally. Seems like there were two main issues:

```
UserWarning: PyOpenCL compiler caching failed with an exception:
[begin exception]
Traceback (most recent call last):
  File "/home/batman/.pyenv/versions/3.11.4/lib/python3.11/site-packages/pyopencl/cache.py", line 495, in create_built_program_from_source_cached
    _create_built_program_from_source_cached(
  File "/home/batman/.pyenv/versions/3.11.4/lib/python3.11/site-packages/pyopencl/cache.py", line 417, in _create_built_program_from_source_cached
    prg.build(options_bytes, [devices[i] for i in to_be_built_indices])
  File "/home/batman/.pyenv/versions/3.11.4/lib/python3.11/site-packages/pyopencl/__init__.py", line 772, in program_build
    compiler_output("%s succeeded, but resulted in non-empty logs:\n%s"
  File "/home/batman/.pyenv/versions/3.11.4/lib/python3.11/site-packages/pyopencl/__init__.py", line 273, in compiler_output
    warn("Non-empty compiler output encountered. Set the "
pyopencl.CompilerWarning: Non-empty compiler output encountered. Set the environment variable PYOPENCL_COMPILER_OUTPUT=1 to see more.
[end exception]
```

and 

```
State:
Ignition: True Engaged: False


State:
Ignition: True Engaged: False
    Waiting for CAN messages...
logcatd proclogd modeld ui soundd locationd calibrationd torqued controlsd deleter navd paramsd plannerd radard thermald statsd
logcatd proclogd modeld ui soundd locationd calibrationd torqued controlsd deleter navd paramsd plannerd radard thermald statsd
Process bridge:
Traceback (most recent call last):
  File "/home/batman/.pyenv/versions/3.11.4/lib/python3.11/multiprocessing/process.py", line 314, in _bootstrap
    self.run()
  File "/home/batman/.pyenv/versions/3.11.4/lib/python3.11/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/dev/shm/openpilot/threepilot/openpilot/tools/sim/bridge/common.py", line 71, in bridge_keep_alive
    self._run(q)
  File "/dev/shm/openpilot/threepilot/openpilot/tools/sim/bridge/common.py", line 197, in _run
    self.print_status()
  File "/dev/shm/openpilot/threepilot/openpilot/tools/sim/bridge/common.py", line 88, in print_status
    print(
BlockingIOError: [Errno 11] write could not complete without blocking

modeld init
setting up CL context
CL context ready; loading model
Onnx available providers:  ['TensorrtExecutionProvider', 'CUDAExecutionProvider', 'AzureExecutionProvider', 'CPUExecutionProvider']
Onnx selected provider:  [('CUDAExecutionProvider', {'cudnn_conv_algo_search': 'DEFAULT'})]
logcatd proclogd modeld ui soundd locationd calibrationd torqued controlsd deleter navd paramsd plannerd radard thermald statsd
Onnx using  ['CUDAExecutionProvider', 'CPUExecutionProvider']
ready to run onnx model {'input_imgs': [1, 12, 128, 256], 'big_input_imgs': [1, 12, 128, 256], 'desire': [1, 100, 8], 'traffic_convention': [1, 2], 'lateral_control_params': [1, 2], 'prev_desired_curv': [1, 100, 1], 'features_buffer': [1, 99, 512]}
models loaded, modeld starting
logcatd proclogd modeld ui soundd locationd calibrationd torqued controlsd deleter navd paramsd plannerd radard thermald statsd
logcatd proclogd modeld ui soundd locationd calibrationd torqued controlsd deleter navd paramsd plannerd radard thermald statsd
logcatd proclogd modeld ui soundd locationd calibrationd torqued controlsd deleter navd paramsd plannerd radard thermald statsd
```

**Review by adeebshihadeh — CHANGES_REQUESTED** [review]
Had no issues running locally now. However, it's easy to make the test pass when it shouldn't with this patch. I think adding a check on distance (based on the time) should be enough. Other than that, seems good to go.

```
diff --git a/tools/sim/bridge/common.py b/tools/sim/bridge/common.py
index 64407143f..997a2e0cf 100644
--- a/tools/sim/bridge/common.py
+++ b/tools/sim/bridge/common.py
@@ -184,6 +184,7 @@ Ignition: {self.simulator_state.ignition} Engaged: {self.simulator_state.is_enga
       brake_out = brake_op if self.simulator_state.is_engaged else brake_manual
       steer_out = steer_op if self.simulator_state.is_engaged else steer_manual

+      throttle_out = 0.
       self.world.apply_controls(steer_out, throttle_out, brake_out)
       self.world.read_state()
       self.world.read_sensors(self.simulator_state)
```

**Comment by bongbui321** [comment]
>  think adding a check on distance (based on the time) should be enough

Was thinking of adding that test for the Metadrive run on GHA since for this one we could still visually see it's moving

**Comment by adeebshihadeh** [comment]
Nice work! Looking forward to getting this in CI.

**Pull request merged**